compiler/ballerina-lang/src/main/java/io/ballerina/projects/CodeGenOptimizationReportEmitter.java

```diff
@@ -0,0 +1,172 @@
+    private static long nativeOptimizationDuration = 0;
+    private static final PrintStream out = System.out;
+    private static final String CODEGEN_OPTIMIZATION_REPORT = "codegen_optimization_report.json";
+
+    protected static void flipBirOptimizationTimer(PackageID pkgId) {
+        if (birOptimizationDurations.containsKey(pkgId)) {
+            long totalDuration = System.currentTimeMillis() - birOptimizationDurations.get(pkgId);
+            birOptimizationDurations.put(pkgId, totalDuration);
+        } else {
+            birOptimizationDurations.put(pkgId, System.currentTimeMillis());
+        }
+    }
+
+    protected static void flipNativeOptimizationTimer() {
+        if (nativeOptimizationDuration == 0) {
+            nativeOptimizationDuration = System.currentTimeMillis();
+        } else {
+            nativeOptimizationDuration = System.currentTimeMillis() - nativeOptimizationDuration;
+        }
+    }
+
+    protected static void emitBirOptimizationDuration() {
+        long totalDuration = birOptimizationDurations.values().stream().mapToLong(Long::longValue).sum();
+        out.printf("Duration for unused BIR node analysis : %dms%n", totalDuration);
+        birOptimizationDurations.forEach((key, value) -> out.printf("    %s : %dms%n", key, value));
+        out.println();
+        // TODO make this class a singleton
+        birOptimizationDurations.clear();
+    }
+
+    protected static void emitNativeOptimizationDuration() {
+        out.println(
+                "Duration for Bytecode Optimization (analysis + deletion) : " + nativeOptimizationDuration + "ms");
+        nativeOptimizationDuration = 0;
+    }
+
+    protected static void emitOptimizedExecutableSize(Path executableFilePath) {
+        try {
+            float optimizedJarSize = Files.size(executableFilePath) / (1024f * 1024f);
+            System.out.printf("Optimized file size : %f MB%n", optimizedJarSize);
+        } catch (IOException e) {
+            throw new ProjectException("Failed to emit optimized executable size ", e);
+        }
+    }
+
+    protected static void emitCodegenOptimizationReport(
+            Map<PackageID, UsedBIRNodeAnalyzer.InvocationData> invocationDataMap, Path projectDirectoryPath,
+            ProjectKind projectKind) {
+        Path reportParentDirectoryPath = projectDirectoryPath.resolve(TARGET_DIR_NAME).toAbsolutePath().normalize();
+
+        if (projectKind == ProjectKind.SINGLE_FILE_PROJECT) {
+            projectDirectoryPath = projectDirectoryPath.toAbsolutePath().getParent();
+            if (projectDirectoryPath != null) {
+                reportParentDirectoryPath = projectDirectoryPath;
+            }
+        }
+
+        if (!Files.exists(reportParentDirectoryPath)) {
+            try {
+                Files.createDirectories(reportParentDirectoryPath);
+            } catch (IOException e) {
+                throw new ProjectException("Failed to create Optimization Report directory ", e);
```

Comment:
Inconsistent case usage.
```suggestion
                throw new ProjectException("Failed to create optimization report directory ", e);
```